onnxruntime/core/session/inference_session.cc

```diff
@@ -1424,32 +1424,30 @@ common::Status InferenceSession::Initialize() {
       if (cuda_ep && cuda_ep->IsGraphCaptureEnabled()) {
-        if (cuda_ep->IsGraphCaptureEnabled()) {
-          if (HasControlflowNodes(graph)) {
-            LOGS(*session_logger_, ERROR) << "This session cannot use the CUDA Graph feature as requested by the user "
-                                          << " as the model has control flow nodes which can't be supported by CUDA Graphs.";
-
-            // Return error status as we don't want the session initialization to complete successfully
-            // if the user has requested usage of CUDA Graph feature and we cannot honor that.
-            ORT_RETURN_IF_ERROR_SESSIONID_(
-                ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
-                                "This session cannot use the CUDA Graph feature as requested by the user "
-                                " as the model has control flow nodes which can't be supported by CUDA Graphs."));
-          } else if (!AreAllNodesInMainGraphAssignedToOneEp(graph, onnxruntime::kCudaExecutionProvider)) {
-            LOGS(*session_logger_, ERROR) << "This session cannot use the CUDA Graph feature as requested by the user "
-                                          << " as all the graph nodes have not been partitioned to the CUDA EP.";
-
-            // Return error status as we don't want the session initialization to complete successfully
-            // if the user has requested usage of CUDA Graph feature and we cannot honor that.
-            ORT_RETURN_IF_ERROR_SESSIONID_(
-                ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
-                                "This session cannot use the CUDA Graph feature as requested by the user "
-                                " as all the graph nodes have not been partitioned to the CUDA EP."));
-
-          } else {
-            LOGS(*session_logger_, INFO) << "This session will use the CUDA Graph feature as requested by the user.";
-            cached_execution_provider_for_graph_replay_.SetExecutionProvider(cuda_ep);
-          }
+        if (HasControlflowNodes(graph)) {
+          LOGS(*session_logger_, ERROR) << "This session cannot use the CUDA Graph feature as requested by the user "
+                                        << " as the model has control flow nodes which can't be supported by CUDA Graphs.";
+
+          // Return error status as we don't want the session initialization to complete successfully
+          // if the user has requested usage of CUDA Graph feature and we cannot honor that.
+          ORT_RETURN_IF_ERROR_SESSIONID_(
+              ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                              "This session cannot use the CUDA Graph feature as requested by the user "
+                              " as the model has control flow nodes which can't be supported by CUDA Graphs."));
+        } else if (!AreAllNodesInMainGraphAssignedToOneEp(graph, onnxruntime::kCudaExecutionProvider)) {
```

Comment:
Many transformer-based models fail on this check, which makes cuda graph could not work here. It seems to be caused by that some embedding lookup related ops are assigned on CPU instead of GPU. Do you know if there are any way to disable this default assignment?

Reply:
Usually, shape related nodes are placed on CPU. It is possible to remove those nodes by "freezing" shape (replace shape related nodes by initializers).

As an example, we have a script that could freeze BERT model:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/shape_optimizer.py
User need specify --input_ids, --segment_ids and --input_mask for 3 graph inputs for BERT model, then use --batch_size and --sequence_length to freeze input shape. Then the script will create dummy inputs to run an inference to know the shape of each intermediate tensor. Finally, it will replace those Shape node output by initializers of actual shape, or replace the shape input of Reshape node by an initializer.

Reply:
Maybe we can think about loosening this gating check to ensure that all **compute** nodes are placed on CUDA EP (i.e.) instead of ensuring ALL nodes are placed on CUDA EP ? We will allow placing the shape nodes on the CPU EP and perform the graph capture. I think this is esentially the same as "freezing" the shape using dummy inputs in an offline script, isn't it ?

Reply:
Yes, that's also what I was thinking about. 

It is feasible while it is probably not a simple task to implement it. 

Those shape computation nodes only need run once, so we actually do not need to run those nodes in CPU again (assuming the shapes do not change).

Reply:
Naively, I thought it is simple to implement this. 

I was thinking - If we see a node assigned to CPU, then traverse all its ascendants and ensure they are all placed all placed on CPU leading upto a `Shape` node && traverse all its children and ensure that they are placed on CPU leading upto a `Reshape` node. We keep marking visited nodes. Keep doing this for unvisited nodes placed on CPU and if we don't see the above condition, then dis-allow using cuda graphs as it contains a **compute** node placed on a non-CUDA EP. I am happy to give this a shot.

I was recently talking to another team who were interested in using cuda graphs for their image models and were wondering if we can avoid the "offline" script to freeze the shape values if possible. For Transformer models, running offline optimization scripts is now part of the workflow and is nothing new, but for teams dealing exclusively with image models, this might be an unnecessary addition to the workflow.

`Those shape computation nodes only need run once, so we actually do not need to run those nodes in CPU again` - Right, but they will only run during the graph capture step - the graph replay doesn't even invoke these nodes.

Reply:
Sure